internal/validate/validate.go

```diff
@@ -148,31 +168,55 @@ func (v *Validator) Validate(ctx *stopper.Context) (*Report, error) {
 	}
 	defer extConn.Drop(ctx, conn)
 
-	stats, err := captureInitialStats(ctx, extConn, conn)
-	if err != nil {
-		return nil, err
-	}
-
-	if err := v.runWorkloadWithBackup(ctx, extConn); err != nil {
-		return nil, err
-	}
-
-	if err := v.runIncrementalBackup(ctx, conn, extConn); err != nil {
-		return nil, err
-	}
-
-	if err := v.checkBackups(ctx, extConn); err != nil {
-		return nil, err
-	}
-
-	if err := v.performRestore(ctx, conn, extConn); err != nil {
-		return nil, err
-	}
-
-	if err := v.verifyIntegrity(ctx, conn); err != nil {
-		// If we fail to verify the integrity, just log the error, but
-		// still provide a complete report
-		slog.Error("failed to verify integrity", slog.Any("error", err))
+	var stats []*db.Stats
+
+	// Define validation steps
+	steps := []validationStep{
+		{
+			name: "capture initial stats",
+			fn: func(ctx *stopper.Context, extConn *db.ExternalConn) error {
+				var err error
+				stats, err = v.captureInitialStats(ctx, extConn)
+				return err
+			},
+		},
+		{
+			name: "workload with backup",
+			fn:   v.runWorkloadWithBackup,
+		},
+		{
+			name: "incremental backup",
+			fn:   v.runIncrementalBackup,
+		},
+		{
+			name: "check backups",
+			fn:   v.checkBackups,
+		},
+		{
+			name: "restore",
+			fn:   v.performRestore,
+		},
+		{
+			name: "verify integrity",
+			fn: func(ctx *stopper.Context, extConn *db.ExternalConn) error {
+				if err := v.verifyIntegrity(ctx); err != nil {
+					// If we fail to verify the integrity, just log the error, but
+					// still provide a complete report
+					slog.Error("failed to verify integrity", slog.Any("error", err))
+				}
+				return nil
+			},
+		},
+	}
+
+	// Execute steps
+	for _, step := range steps {
```

Comment:
Is a full state machine really needed here?  It's fine I guess, just seems over-complicated.  And is begging for a test.  You could just repeat a check for is stopping between the calls.  Anyway, up to you.

Reply:
I had the check repeated originally, I decided to refactor it to make the steps more declarative. 
I guess it's just matter of preference. As per the testing, it's covered in the minio integration test.